src/indexer/index_sorter.rs

```diff
@@ -0,0 +1,426 @@
+use super::SegmentWriter;
+use crate::{DocId, IndexSettings, Order, TantivyError};
+use std::cmp::Reverse;
+
+/// Struct to provide mapping from old docid to new docid and vice versa
+pub struct DocidMapping {
+    new_docid_to_old: Vec<u32>,
+    old_docid_to_new: Vec<u32>,
+}
+
+impl DocidMapping {
+    /// returns the new docid for the old docid
+    pub fn get_new_docid(&self, docid: DocId) -> DocId {
+        self.old_docid_to_new[docid as usize]
+    }
+    /// returns the old docid for the new docid
+    pub fn get_old_docid(&self, docid: DocId) -> DocId {
+        self.new_docid_to_old[docid as usize]
+    }
+    /// iterate over old docids in order of the new docids
+    pub fn iter_old_docids(&self) -> std::slice::Iter<'_, u32> {
+        self.new_docid_to_old.iter()
+    }
+}
+
+pub(crate) fn sort_index(
+    settings: IndexSettings,
+    segment_writer: &mut SegmentWriter,
+) -> crate::Result<DocidMapping> {
+    let docid_mapping = get_docid_mapping(settings, segment_writer)?;
+    Ok(docid_mapping)
+}
+
+// Generates a document mapping in the form of [index new docid] -> old docid
+fn get_docid_mapping(
+    settings: IndexSettings,
+    segment_writer: &mut SegmentWriter,
+) -> crate::Result<DocidMapping> {
+    let field_id = segment_writer
+        .segment_serializer
+        .segment()
+        .schema()
+        .get_field(&settings.sort_by_field.field)
+        .ok_or_else(|| {
+            TantivyError::InvalidArgument(format!(
+                "field to sort index by not found: {:?}",
+                settings.sort_by_field.field
+            ))
+        })?;
+    // for now expect fastfield, but not strictly required
+    let fast_field = segment_writer
+        .fast_field_writers
+        .get_field_writer(field_id)
+        .ok_or_else(|| {
+            TantivyError::InvalidArgument(format!(
+                "sort index by field is required to be a fast field {:?}",
+                settings.sort_by_field.field
+            ))
+        })?;
+
+    // create new docid to old docid index (used in fast_field_writers)
+    let data = fast_field.get_data();
+    let mut docid_and_data = data
+        .into_iter()
+        .enumerate()
+        .map(|el| (el.0 as DocId, el.1))
+        .collect::<Vec<_>>();
+    if settings.sort_by_field.order == Order::Desc {
+        docid_and_data.sort_unstable_by_key(|k| Reverse(k.1));
```

Comment:
i suspect stable sort might be useful here.

Reply:
yep, done